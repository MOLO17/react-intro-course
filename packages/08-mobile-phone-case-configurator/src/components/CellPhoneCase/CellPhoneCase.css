div.cell-phone-case {
  display: flex;
  position: relative;
  width: 320px;
  height: 576px;
  border-radius: 48px;
  box-shadow: 0 16px 32px rgba(0, 0, 0, 0.2);
  align-items: center;
  justify-content: center;
}

div.cell-phone-case::before {
  position: absolute;
  display: block;
  content: '';
  background-color: black;
  width: 48px;
  height: 128px;
  border-radius: 24px;
  right: 24px;
  top: 24px;
}

div.cell-phone-case > span {
  writing-mode: vertical-lr;
  font-weight: bold;
  font-size: 40px;
  line-height: 72px;
  padding: 16px 0;
  text-transform: uppercase;
  border-radius: 36px;
}
