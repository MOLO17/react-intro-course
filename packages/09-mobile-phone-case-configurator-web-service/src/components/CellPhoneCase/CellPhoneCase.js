import React from 'react';

import Color from 'color';

import PropTypes from 'prop-types';

import checkeredTexture from './textures/Checkered.svg';
import stripedTexture from './textures/Striped.svg';
import dottedTexture from './textures/Dotted.svg';

import './CellPhoneCase.css';

const TEXTURE_MAP = {
  DOTTED: dottedTexture,
  STRIPED: stripedTexture,
  CHECKERED: checkeredTexture,
};

const CellPhoneCase = ({ color, text, texture }) => (
  <div
    className="cell-phone-case"
    style={{
      backgroundColor: color,
      ...(typeof TEXTURE_MAP[texture] === 'string'
        ? { backgroundImage: `url('${TEXTURE_MAP[texture]}')` }
        : {}),
    }}
  >
    <span
      style={{
        backgroundColor: color,
        color: new Color(color).isDark() ? 'white' : 'black',
      }}
    >
      {text}
    </span>
  </div>
);

CellPhoneCase.propTypes = {
  color: PropTypes.string.isRequired,
  text: PropTypes.string,
  texture: PropTypes.oneOf(['NONE', 'DOTTED', 'STRIPED', 'CHECKERED']),
};

CellPhoneCase.defaultProps = {
  text: '',
  texture: undefined,
};

export default CellPhoneCase;
