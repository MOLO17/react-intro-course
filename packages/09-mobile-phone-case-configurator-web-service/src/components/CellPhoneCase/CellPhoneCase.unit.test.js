import React from 'react';
import renderer from 'react-test-renderer';

import CellPhoneCase from './CellPhoneCase';

describe('CellPhoneCase component', () => {
  it('renders correctly', () => {
    const component = renderer.create(
      <CellPhoneCase color="#ff0000" texture="DOTTED" text="Dummy text" />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe('CellPhoneCase color', () => {
  it('should be displayed correctly', () => {
    const color = '#ff0000';

    const component = renderer.create(<CellPhoneCase color={color} />);

    const caseBody = component.root.findByProps({
      className: 'cell-phone-case',
    });

    const { backgroundColor } = caseBody.props.style;

    expect(backgroundColor).toEqual(color);
  });
});

describe('CellPhoneCase texture', () => {
  describe('dotted texture', () => {
    it('should be displayed correctly', () => {
      const component = renderer.create(
        <CellPhoneCase color="#ff0000" texture="DOTTED" />,
      );

      const caseBody = component.root.findByProps({
        className: 'cell-phone-case',
      });

      const { backgroundImage } = caseBody.props.style;

      expect(backgroundImage).toEqual(`url('Dotted.svg')`);
    });
  });

  describe('checkered texture', () => {
    it('should be displayed correctly', () => {
      const component = renderer.create(
        <CellPhoneCase color="#ff0000" texture="CHECKERED" />,
      );

      const caseBody = component.root.findByProps({
        className: 'cell-phone-case',
      });

      const { backgroundImage } = caseBody.props.style;

      expect(backgroundImage).toEqual(`url('Checkered.svg')`);
    });
  });

  describe('striped texture', () => {
    it('should be displayed correctly', () => {
      const component = renderer.create(
        <CellPhoneCase color="#ff0000" texture="STRIPED" />,
      );

      const caseBody = component.root.findByProps({
        className: 'cell-phone-case',
      });

      const { backgroundImage } = caseBody.props.style;

      expect(backgroundImage).toEqual(`url('Striped.svg')`);
    });
  });

  describe('plain texture', () => {
    it('should be displayed correctly', () => {
      const component = renderer.create(
        <CellPhoneCase color="#ff0000" texture="NONE" />,
      );

      const caseBody = component.root.findByProps({
        className: 'cell-phone-case',
      });

      const { backgroundImage } = caseBody.props.style;

      expect(backgroundImage).toBe(undefined);
    });
  });

  describe('undefined texture', () => {
    it('should be displayed correctly', () => {
      const component = renderer.create(<CellPhoneCase color="#ff0000" />);

      const caseBody = component.root.findByProps({
        className: 'cell-phone-case',
      });

      const { backgroundImage } = caseBody.props.style;

      expect(backgroundImage).toBe(undefined);
    });
  });
});

describe('CellPhoneCase text', () => {
  it('should be inserted correctly', () => {
    const text = 'Dummy text';

    const component = renderer.create(
      <CellPhoneCase color="#ff0000" text={text} />,
    );

    const textNode = component.root.findByProps({
      children: text,
    });

    expect(textNode).toBeTruthy();
  });

  it('should be white if color is dark', () => {
    const text = 'Dummy text';

    const component = renderer.create(
      <CellPhoneCase color="#000000" text={text} />,
    );

    const textNode = component.root.findByProps({
      children: text,
    });

    const { color } = textNode.props.style;

    expect(color).toEqual('white');
  });

  it('should be black if color is light', () => {
    const text = 'Dummy text';

    const component = renderer.create(
      <CellPhoneCase color="#ffffff" text={text} />,
    );

    const textNode = component.root.findByProps({
      children: text,
    });

    const { color } = textNode.props.style;

    expect(color).toEqual('black');
  });
});
