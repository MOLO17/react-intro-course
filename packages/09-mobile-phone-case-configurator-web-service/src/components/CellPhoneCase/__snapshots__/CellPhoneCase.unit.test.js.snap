// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CellPhoneCase component renders correctly 1`] = `
<div
  className="cell-phone-case"
  style={
    Object {
      "backgroundColor": "#ff0000",
      "backgroundImage": "url('Dotted.svg')",
    }
  }
>
  <span
    style={
      Object {
        "backgroundColor": "#ff0000",
        "color": "white",
      }
    }
  >
    Dummy text
  </span>
</div>
`;
